### PR TITLE
Ability to add engine size for particular step in batch config

### PR DESCRIPTION
**Pull request by AndrewNepogoda**
Workflow resource manager can define multiple engines for execution and provide rai config for for steps which required particular engine size.
The engine which add as a first to managed engines marked as default and can be deleted only during resources clean up process.

### CODE REVIEW

File: workflow/executor.py [hunk 1]
```diff
@@ -410,7 +415,12 @@ def run(self):
                               q.update_step_state(step.idt, WorkflowStepState.IN_PROGRESS.name), readonly=False,
                               ignore_problems=True)
             try:
-                step.execute(self.logger, self.config.env, rai_config)
+                if step.engine_size:
```
Comment (stanislav-bedunkevich) [inline workflow/executor.py:418]: so if I specify `xs` as the default and one of my steps happens to have `xs` as well, line 421 would throw? I'd expect it to not throw anything and just proceed with the default engine.

File: workflow/manager.py [hunk 2]
```diff
@@ -6,24 +9,42 @@
...
+    __logger: logging.Logger
+    __rai_config: RaiConfig
+    __engines: dict[str, EngineMetaInfo]
 
     def __init__(self, logger: logging.Logger, rai_config: RaiConfig):
-        self.logger = logger
-        self.rai_config = rai_config
+        self.__logger = logger
+        self.__rai_config = rai_config
+        self.__engines = {}
```
Comment (stanislav-bedunkevich) [inline workflow/manager.py:27]: would at any point here be more than two engines (default and a worker needed for a step which has other engine spec) active?

**Review by stanislav-bedunkevich — CHANGES_REQUESTED** [review]
Looks good to me, left a few comments though. The only change I think is needed is the ability to just run silently any steps that have the same engine size as the current default, without raising any exceptions or deleting it.

File: workflow/executor.py (hunk 1 above)
Comment (stanislav-bedunkevich) [reply workflow/executor.py:418]: might be also good to "check" what is needed for the next step, so if it happens to run on the same size as just provisioned we don't delete & reprovision a new one

File: workflow/manager.py (hunk 2 above)
Comment (AndrewNepogoda) [reply workflow/manager.py:27]: not now. But it might be requested to keep engine after step execution. Then it won't require changes in manager logic.

**Review by stanislav-bedunkevich — APPROVED** [review]
Thanks!

**Pull request merged**